karpenter-operator/controllers/karpenter/machine_approver.go

```diff
@@ -0,0 +1,232 @@
+)
+
+type MachineApproverController struct {
+	client     client.Client
+	certClient *certificatesv1client.CertificatesV1Client
+}
+
+func (r *MachineApproverController) SetupWithManager(mgr ctrl.Manager) error {
+	certClient, err := certificatesv1client.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+	r.certClient = certClient
+	r.client = mgr.GetClient()
+
+	c, err := controller.New("karpenter_machine_approver", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return fmt.Errorf("failed to construct karpenter_machine_approver controller: %w", err)
+	}
+
+	csrFilterFn := func(csr *certificatesv1.CertificateSigningRequest) bool {
+		if csr.Spec.SignerName != certificatesv1.KubeAPIServerClientKubeletSignerName {
+			return false
+		}
+		// only reconcile pending CSRs (not approved and not denied).
+		if !certificates.IsCertificateRequestPending(csr) {
+			return false
+		}
+		// only reconcile kubernetes.io/kube-apiserver-client-kubelet when it is created by the node bootstrapper
+		if csr.Spec.Username != nodeBootstrapperUsername {
+			mgr.GetLogger().Info("Ignoring csr because it is not from the node bootstrapper", "csr", csr.Name)
+			return false
+		}
+		return true
+	}
+
+	if err := c.Watch(source.Kind(
+		mgr.GetCache(),
+		&certificatesv1.CertificateSigningRequest{},
+		&handler.TypedEnqueueRequestForObject[*certificatesv1.CertificateSigningRequest]{},
+		predicate.NewTypedPredicateFuncs(csrFilterFn),
+	)); err != nil {
+		return fmt.Errorf("failed to watch CertificateSigningRequest: %v", err)
+	}
+
+	return nil
+}
+
+func (r *MachineApproverController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling CSR", "req", req)
+
+	csr := &certificatesv1.CertificateSigningRequest{}
+	if err := r.client.Get(ctx, req.NamespacedName, csr); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get csr %s: %v", req.NamespacedName, err)
+	}
+
+	// Return early if deleted
+	if !csr.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// If a CSR is approved/denied after being added to the queue,
+	// but before we reconcile it, trying to approve it will result in an error and cause a loop.
+	// Return early if the CSR has been approved/denied externally.
+	if !certificates.IsCertificateRequestPending(csr) {
+		log.Info("CSR is already processed ", "csr", csr.Name)
+		return ctrl.Result{}, nil
+	}
+
+	approved, err := r.authorize(ctx, csr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if approved {
+		log.Info("Attempting to approve CSR", "csr", csr.Name)
+		if err := r.approve(ctx, csr); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to approve csr %s: %v", csr.Name, err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MachineApproverController) authorize(ctx context.Context, csr *certificatesv1.CertificateSigningRequest) (bool, error) {
```

Comment:
can we add a todo to include a creation time window for the nodeclaim, the instance and csr triplets and also ratelimit and short circuit approval based on the number of pending CSRs?

Reply:
done